cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -72,32 +99,48 @@ func (pe *ProbeEvent) addBlockDeviceEvent(msg controller.EventMessage) {
 
 // deleteBlockDeviceEvent deactivate blockdevice resource using uuid from etcd
 func (pe *ProbeEvent) deleteBlockDeviceEvent(msg controller.EventMessage) {
-	blockDeviceOk := pe.deleteBlockDevice(msg)
-
-	// when one disk is removed from node and entry related to
-	// that disk is not present in etcd,  in that case it
-	// again rescan full system and update etcd accordingly.
-	if !blockDeviceOk {
-		go pe.initOrErrorEvent()
-	}
-}
-
-func (pe *ProbeEvent) deleteBlockDevice(msg controller.EventMessage) bool {
-	bdList, err := pe.Controller.ListBlockDeviceResource()
+	bdAPIList, err := pe.Controller.ListBlockDeviceResource(false)
 	if err != nil {
 		klog.Error(err)
-		return false
 	}
-	ok := true
-	for _, diskDetails := range msg.Devices {
-		existingBlockDeviceResource := pe.Controller.GetExistingBlockDeviceResource(bdList, diskDetails.UUID)
+
+	isDeactivated := true
+	isGPTBasedUUIDEnabled := pe.Controller.FeatureGates.IsEnabled(features.GPTBasedUUID)
+
+	for _, device := range msg.Devices {
+		// create the new UUID for removing the device
+		if isGPTBasedUUIDEnabled {
+			_, ok := pe.Controller.BDHierarchy[device.DevPath]
+			if !ok {
+				klog.Infof("Disk %s not in hierarchy", device.DevPath)
+				// not in hierarchy continue
+				continue
+			}
+			// remove from the hierarchy
+			delete(pe.Controller.BDHierarchy, device.DevPath)
```

Comment:
the block device state -- is marked as inactive, but it is still present in the etcd. Will deleting from the cache possibly cause any issues with later processing. 

Can you explain what happens in the following case:

=> A detach of a device with partition occurs. Delete events will trigger. 
=> Attach event of the same device with partition occurs.

Reply:
> the block device state -- is marked as inactive, but it is still present in the etcd. Will deleting from the cache possibly cause any issues with later processing.

No. When the device gets reconnected to the node, it will be first added to the cache, and then marked as Active.

Reply:
> => A detach of a device with partition occurs. Delete events will trigger.

Suppose the disk has 2 partitions `sda1` and `sda2`.
Sequence of events when device `sda` is disconnected
1. 3 udev `remove` events will be generated in the following order.
```    
remove sda1
remove sda2
remove sda
```
2. When remove event  for `sda1` and `sda2` is processed, we generate the UUID for `sda1` and `sda2`, and deactivate the resource. If the resource is not in etcd it is skipped. (for cstor created partitions it can be this case)
3. When remove event for `sda` is processed, we generate the UUID for `sda` and deactivate the resource.
NOTE: If UUID cannot be generated, we simple skip the event, because the add event gurantees that all devices will have a UUID (either by creating partition or by using WWN)


> => Attach event of the same device with partition occurs.

The above same disk is attached back. 3 udev `add` events will be generated in the following order
```
add sda
add sda1
add sda2
```
Steps for processing:
1. `udev add for sda`: We try to generate UUID for the device. If it fails, we check if there are partitions on the device, then we will skip the event.
2. `udev add for sda1/sda2`: generate UUID for the device. Check if device is claimed. If the device is claimed, update only the path, capacity, node attributes and state. If the device is not claimed, update all the details on the blockdevice resources.

Reply:
Could this be dependent on the kernel or is the order guaranteed?